core/src/main/java/org/apache/accumulo/core/file/FileOperations.java

```diff
@@ -48,38 +48,295 @@ public static FileOperations getInstance() {
+  /**
+   * Construct an operation object allowing one to create a "scan" reader for a file. Scan readers do not have any optimizations for seeking beyond their
+   * initial position. This is useful for file operations that only need to scan data within a range and do not need to seek. Therefore file metadata such as
+   * indexes does not need to be kept in memory while the file is scanned. Also seek optimizations like bloom filters do not need to be loaded. <br>
+   * Syntax:
+   *
+   * <pre>
+   * FileSKVIterator scanner = fileOperations.openScanReader()
+   *     .ofFile(...)
+   *     .overRange(...)
+   *     .withTableConfiguration(...)
+   *     .withRateLimiter(...) // optional
+   *     .withBlockCache(...) // optional
+   *     .execute();
+   * </pre>
+   */
+  public OpenScanReaderOperation openScanReader() {
+    return new OpenScanReaderOperation();
+  }
 
   /**
-   * Open a reader that fully support seeking and also enable any optimizations related to seeking, like bloom filters.
+   * Construct an operation object allowing one to create a reader for a file. A reader constructed in this manner fully supports seeking, and also enables any
+   * optimizations related to seeking (e.g. Bloom filters). <br>
+   * Syntax:
    *
+   * <pre>
+   * FileSKVIterator scanner = fileOperations.openReader()
+   *     .ofFile(...)
+   *     .withTableConfiguration(...)
+   *     .withRateLimiter(...) // optional
+   *     .withBlockCache(...) // optional
+   *     .seekToBeginning(...) // optional
+   *     .execute();
+   * </pre>
+   */
+  public OpenReaderOperation openReader() {
+    return new OpenReaderOperation();
+  }
+
+  //
+  // Operation objects.
+  //
+
+  /**
+   * Options common to all FileOperations.
+   */
+  protected static class FileAccessOperation<SubclassType extends FileAccessOperation<SubclassType>> {
+    private AccumuloConfiguration tableConfiguration;
+
+    private String filename;
+    private FileSystem fs;
+    private Configuration fsConf;
+
+    /** Specify the table configuration defining access to this file. */
+    @SuppressWarnings("unchecked")
+    public SubclassType withTableConfiguration(AccumuloConfiguration tableConfiguration) {
+      this.tableConfiguration = tableConfiguration;
+      return (SubclassType) this;
+    }
+
+    /** Specify the file this operation should apply to. */
+    @SuppressWarnings("unchecked")
+    public SubclassType ofFile(String filename, FileSystem fs, Configuration fsConf) {
+      this.filename = filename;
+      this.fs = fs;
+      this.fsConf = fsConf;
+      return (SubclassType) this;
+    }
+
+    public String getFilename() {
+      return filename;
+    }
+
+    public FileSystem getFileSystem() {
+      return fs;
+    }
+
+    public Configuration getConfiguration() {
+      return fsConf;
+    }
+
+    public AccumuloConfiguration getTableConfiguration() {
```

Comment:
I didn't, but I'll consider it.  There's a balance here.  Keeping to a purely 'fluent' style, I should have a function call for each separate parameter, e.g.

```java
openWriter()
    .ofFile(...)
    .onFileSystem(...)
    .withFileSystemConfiguration(conf)
    .withTableConfiguration(...)
    ...
```
I decided that 'file', and 'filesystem' were reasonable to lump together to cut down on verbosity. I'm not entirely sure where the `Configuration` should go, particularly as something like 50% of callers just use the equivalent of `fs.getConf()` for this, but I lumped it together with 'file' too.  On the other hand, table configuration seems to me to really be a very different concept.

Perhaps (?):
```java
openWriter()
.ofFile(filename, fs)
.withConfiguration(conf, tableConf)
...
```